core/drivers/crypto/aspeed/hace_ast2600.c

```diff
@@ -0,0 +1,376 @@
+		if (rc)
+			return rc;
+
+		rc = ast2600_hace_update(ctx, (uint8_t *)&dbits_be[0],
+					 sizeof(dbits_be[0]));
+		if (rc)
+			return rc;
+		break;
+	case TEE_ALG_SHA384:
+	case TEE_ALG_SHA512:
+		if (last < 112)
+			padn = 112 - last;
+		else
+			padn = 240 - last;
+
+		rc = ast2600_hace_update(ctx, pad, padn);
+		if (rc)
+			return rc;
+
+		rc = ast2600_hace_update(ctx, (uint8_t *)&dbits_be[1],
+					 sizeof(dbits_be[1]));
+		if (rc)
+			return rc;
+
+		rc = ast2600_hace_update(ctx, (uint8_t *)&dbits_be[0],
+					 sizeof(dbits_be[0]));
+		if (rc)
+			return rc;
+		break;
+	default:
+		return TEE_ERROR_NOT_SUPPORTED;
+	}
+
+	cache_operation(TEE_CACHEINVALIDATE, hctx->digest,
+			sizeof(hctx->digest));
+
+	memcpy(digest, hctx->digest, hctx->dgt_size);
+
+	return TEE_SUCCESS;
+}
+
+static void ast2600_hace_free(struct crypto_hash_ctx *ctx)
+{
+	struct ast2600_hace_ctx *hctx = NULL;
+
+	hctx = container_of(ctx, struct ast2600_hace_ctx, hash_ctx);
+
+	free(hctx);
+}
+
+static void ast2600_hace_copy_state(struct crypto_hash_ctx *dst_ctx,
+				    struct crypto_hash_ctx *src_ctx)
+{
+	struct ast2600_hace_ctx *src_hctx = NULL;
+	struct ast2600_hace_ctx *dst_hctx = NULL;
+
+	src_hctx = container_of(src_ctx, struct ast2600_hace_ctx, hash_ctx);
+	dst_hctx = container_of(dst_ctx, struct ast2600_hace_ctx, hash_ctx);
+
+	cache_operation(TEE_CACHEINVALIDATE, src_hctx, sizeof(*src_hctx);
+
+	memcpy(dst_hctx, src_hctx, sizeof(*dst_hctx));
+
+	cache_operation(TEE_CACHEFLUSH,	dst_hctx, sizeof(*dst_hctx));
+}
+
+static const struct crypto_hash_ops ast2600_hace_ops = {
+	.init = ast2600_hace_init,
+	.update = ast2600_hace_update,
+	.final = ast2600_hace_final,
+	.free_ctx = ast2600_hace_free,
+	.copy_state = ast2600_hace_copy_state,
+};
+
+static TEE_Result ast2600_hace_alloc(struct crypto_hash_ctx **pctx,
+				     uint32_t algo)
+{
+	struct ast2600_hace_ctx *hctx;
+
+	hctx = calloc(1, sizeof(*hctx));
+	if (!hctx)
+		return TEE_ERROR_OUT_OF_MEMORY;
+
+	hace_ctx->ops = &ast2600_hace_ops;
+	hace_ctx->algo = algo;
+	hace_ctx->cmd = HACE_HASH_CMD_ACCUM | HACE_HASH_CMD_SHA_BE;
+
+	switch (algo) {
+	case TEE_ALG_SHA1:
+		hctx->dgt_size = 20;
+		hctx->blk_size = 64;
+		hctx->pad_size = 8;
+		hctx->cmd |= HACE_HASH_CMD_ALG_SHA1;
+		break;
+	case TEE_ALG_SHA256:
+		hctx->dgt_size = 32;
+		hctx->blk_size = 64;
+		hctx->pad_size = 8;
+		hctx->cmd |= HACE_HASH_CMD_ALG_SHA256;
+		break;
+	case TEE_ALG_SHA384:
+		hctx->dgt_size = 48;
+		hctx->blk_size = 128;
+		hctx->pad_size = 16;
+		hctx->cmd |= HACE_HASH_CMD_ALG_SHA384;
+		break;
+	case TEE_ALG_SHA512:
+		hctx->dgt_size = 64;
+		hctx->blk_size = 128;
+		hctx->pad_size = 16;
+		hctx->cmd |= HACE_HASH_CMD_ALG_SHA512;
+		break;
+	default:
+		free(hctx);
+		return TEE_ERROR_NOT_IMPLEMENTED;
+	}
+
+	*pctx = &hctx->hash_ctx;
+
+	return TEE_SUCCESS;
+}
+
+TEE_Result ast2600_drvcrypt_register_hash(void)
+{
+	hace_virt = core_mmu_get_va(HACE_BASE, MEM_AREA_IO_SEC,
+				    SMALL_PAGE_SIZE);
```

Comment:
Align `SMALL_PAGE_SIZE` under `HACE_BASE`

Reply:
In VIM with tabstop=8, `SMALL_PAGE_SIZE` seems to be aligned under `HACE_BASE`.
Did I miss something? Thanks.

Reply:
The fix commits are squashed for review convenience.